### ISSUE TITLE
TFGAN ideas for pretraining/training

### ISSUE THREAD

**Issue by sleighsoft** [tensorflow/tensorflow#15271]
With the PR #14723 enabling `get_hooks_fn` to be set manually instead of the default `1` step generator and `1` step discriminator there comes a set of new "problems" / things to consider.

1. The `Estimator` saves configured tensor summaries in the background. This does not work when doing something like a `2/2` split for generator/discriminator or a `2/3`. Then `Estimator` will only save one step instead of the actual amounts of steps taken (right?). This means that we have to consider an option to configure the `FileWriter`. I believe that is possible for the vanilla `Estimator` with `scaffolds` but not for the `GANEstimator` currently. This `FileWriter` has to be accessible by the `RunTrainOpsHook`. We would also need a generator + discriminator specific "global_step" that will be used in the `RunTrainOpsHook`. A quick idea would be to use the `overall_global_step * train_steps` of a `RunTrainOpsHook` or to use `dummy_global_step_generator` and `dummy_global_step_discriminator`.

2. When "pre-training" (with a normal call to `.train()` and a modified `sequential_train_hook(10,10)`) the generator for e.g. `10` steps and then the discriminator for `10` steps. Does the discriminator `loss_fn` receive 10 times new data from the generator through `gan_model.discriminator_real_outputs` or will it always be the same data? For pre-training I would assume I can feed in the same output data from the generator in batches multiple epochs. I believe that is not possible in the current setup, but correct me if I am wrong.

3. I have different loss functions for both pre-training and training. There is not `ModeKeys.PRETRAIN` to switch between them.

If I find more things I'll add them here.

**Comment by joel-shor** [comment]
1) The GANEstimator is purposefully designed to record 1 step for 2/2 if that is what's specified. If want it to be saved as 2 steps, you should specify 1/1, no?

2) The generator would get 10 different batches of data. That is the normal behavior; AFAIK it is not standard to train multiple times on the same minibatch.

3) That behavior isn't standard in any estimator. An easy fix would be to have your loss function inside a `tf.cond` based on the global step.

**Comment by sleighsoft** [comment]
1. It makes it difficult to compare the performance of runs with different splits. If there is an easy way to configure it I would use it but if that is too much hassle to implement I can also live without it.
2. You are so correct, I completely misinterpreted that ....
3. The idea with `tf.cond` is great.

**Comment by sleighsoft** [comment]
Is it ok if I open up a pull request that allows to disable `_use_check_shapes` in `GANEstimator`?

**Comment by joel-shor** [comment]
I'm waiting on a review to make this change internally, then it will be
pushed to github within 24 hours. Sorry for the delay.

**Comment by sleighsoft** [comment]
Great :)

**Comment by sleighsoft** [comment]
An idea regarding pre-training
```
from tensorflow.contrib.gan.python import namedtuples
from tensorflow.python.training import session_run_hook
from tensorflow.python.training import training_util
import tensorflow as tf


class RunTrainOpsHook(session_run_hook.SessionRunHook):
  """A hook to run train ops a fixed number of times."""

  def __init__(self, train_ops, pretrain_steps, train_steps):
    """Run train ops a certain number of times.

    Args:
      train_ops: A train op or iterable of train ops to run.
      train_steps: The number of times to run the op(s).
    """
    if not isinstance(train_ops, (list, tuple)):
      train_ops = [train_ops]
    self._pretrain_steps = pretrain_steps
    self._train_ops = train_ops
    self._train_steps = train_steps

  def before_run(self, run_context):
    global_step = training_util.get_or_create_global_step()
    global_step_value = tf.train.global_step(run_context.session, global_step)
    steps = self._pretrain_steps
    if global_step_value > 0:
      steps = self._train_steps
    for _ in range(steps):
      run_context.session.run(self._train_ops)


def get_sequential_train_hooks(pretrain_steps=namedtuples.GANTrainSteps(10, 1),
                               train_steps=namedtuples.GANTrainSteps(1, 1)):
  """Returns a hooks function for sequential GAN training.

  Args:
    train_steps: A `GANTrainSteps` tuple that determines how many generator
      and discriminator training steps to take.

  Returns:
    A function that takes a GANTrainOps tuple and returns a list of hooks.
  """
  def get_hooks(train_ops):

    generator_hook = RunTrainOpsHook(train_ops.generator_train_op,
                                     pretrain_steps.generator_train_steps,
                                     train_steps.generator_train_steps)
    discriminator_hook = RunTrainOpsHook(
        train_ops.discriminator_train_op,
        pretrain_steps.discriminator_train_steps,
        train_steps.discriminator_train_steps)
    return [generator_hook, discriminator_hook]
  return get_hooks
```
The `RunTrainOpsHook` allows to set the number of pretraining steps which will be used only for the first step. After that you can configure the actual adverserial training approach e.g. 1/1

**Comment by sleighsoft** [comment]
Alternatively a new `GANTrainSteps` class can be created or the current one could be extended with pretrain steps to make it a 4 tuple `((pretrain_gen, pretrain_dis), (train_gen, train_dis))`.

**Comment by sleighsoft** [comment]
I looked at the graph of `GANEstimator` in Tensorboard and I have an additional `Discriminator_1`.
Is that to be expected?

![image](https://user-images.githubusercontent.com/9438971/33876870-026f8cda-df27-11e7-86d1-2766f21574f1.png)

**Comment by joel-shor** [comment]
1) One discriminator is for the real data, the other is for the fake data. The weights are shared.
2) To support pretraining, how would you feel about making a function in `train.py` called `get_pretraining_sequential_train_steps(pretraining_steps, train_steps)` that runs `pretraining.discriminator_train_steps` pretraining for D, `pretraining.generator_train_steps` pretraining for G, then `train_steps` alternating?

**Comment by sleighsoft** [comment]
Where is `get_sequential_train_steps` used right now?

**Comment by joel-shor** [comment]
1) That is the normal variable scope naming mechanism. You could add an extra encapsulating name scope
2) A few places. You should use code search to find them all

**Comment by sleighsoft** [comment]
When searching in `tensorflow\contrib\gan\python` I only find it to be defined in `contrib\gan\python\train.py` and never used.

**Comment by sleighsoft** [comment]
Regarding 1. in your comment https://github.com/tensorflow/tensorflow/issues/15271#issuecomment-350702852

When altering the train_steps or get_hooks_fn, how would I log the loss for the pretraining steps?

**Comment by sleighsoft** [comment]
Is there a way to do gradient clipping with `GANEstimator` already?

**Comment by joel-shor** [comment]
https://github.com/tensorflow/models/blob/master/research/gan/pix2pix/train.py#L123

**Comment by sleighsoft** [comment]
Regarding 3. in your https://github.com/tensorflow/tensorflow/issues/15271#issuecomment-350702852

The `global_step` is currently updated non-deterministically so it can happen that while it still runs training the global step already gets updated (not tested for loss calcluation but happens for `tf.cond(global_step,...)` in `generator_fn`). Maybe `global_step_inc` should depend on the loss with ` tf.control_dependencies `?

**Comment by sleighsoft** [comment]
Is transform_grads_fn already exposed in ganestimator? I can't seem to find it.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by joel-shor** [comment]
Sorry for the delay.

I'm open to threading `transform_grads_fn` through the GANEstimator. Would you like to send a PR with that?

You should be able to use normal debugging techniques to check that your pretraining `get_hooks_fn` is correct

**Comment by ptats** [comment]
Not sure if this comment belongs here, but I am wondering about performing weight clipping in the discriminator when using GANEstimator, is this possible?

**Comment by ptats** [comment]
@joel-shor has threading for `transform_grads_fn` been added to the GANEstimator yet?
